### ISSUE TITLE
Consider creating new macro for asserting without redundancy

### ISSUE THREAD

**Issue by magurotuna** [denoland/deno_lint#425]
See https://github.com/denoland/deno_lint/blob/4175773da7a43daecb80bbb5cd3a9025279321bb/src/rules/ban_ts_comment.rs#L174-L200

This test has three cases, and their expected messages & hints are identical.
In such cases, it would be nice if we could removed redundancy i.e. we need to specify expected message and hint _only once_ instead of specifying multiple times.

To make it realize, we should consider a new macro interface first, and then implement it.

**Comment by subhakundu** [comment]
Is there anyone working on it? If not, I can pick this up. Is it okay? It seems a good opportunity for me to learn things.

**Comment by magurotuna** [comment]
@subhakundu Please go ahead! Before you get started to implement, consider a good & ergonomic macro interface, because macros tend to be significantly complicated due to the too high power of expression.
Feel free to ask me (and also other developers) if you have any questions! I'm available here and on discord.

**Comment by subhakundu** [comment]
Thank you @magurotuna, I appreciate it.

**Comment by subhakundu** [comment]
@magurotuna, with this [commit](https://github.com/denoland/deno_lint/commit/b526e5a6af475a9fba29277a1b744ec2f7596d15), the above-referred code have been modified with DirectiveKind enum. The repeated texts are being captured in static functions. In this commit, we have captured the texts in a single place, but following code block

{
  col: 0,
  message: DirectiveKind::ExpectError.as_message(),
  hint: DirectiveKind::ExpectError.as_hint(),
}

(and another two) are being repeated multiple times. We can capture these in simpler macro (or function) and pass ExpectError, Ignore, and Nocheck as arguments. We can create util, so that it can be reused at other places. I think you indicated a similar kind of approach. Please let me know if you think otherwise. I will be more than happy to incorporate your feedback.

Note: Please forgive me for mistaken Rust's vocabulary. I have just started learning.

**Comment by magurotuna** [comment]
@subhakundu Yes, some rules use `enum` for their message and hint, and others don't. (for example [no-await-in-loop](https://github.com/denoland/deno_lint/blob/0827bc1a3ee800d9d714072ba754da2b3e8717b1/src/rules/no_await_in_loop.rs#L10-L11))

Your idea sounds good but also a little vague to me. What I think naively is specifically as follows:

```rust
new_macro! {
  RuleName,
  "message to be asserted",
  "hint to be asserted",
  "<code>":  [
    {
      col: 3
      line: 10,
    },
  ],
};
```

This macro would "group" test cases by messages and hints so that repetitive specifications of the same message and hint could be avoided. But this is just my idea which is very naive and we'd have to consider how to deal with `enum` message and hint. If you come up with a better solution, let's go with it.

**Comment by subhakundu** [comment]
@magurotuna, I dived deep into your comments and tried an [implementation](https://github.com/subhakundu/deno_lint/commit/cb6cf86ce1d17db40a18219007c38753db40135b) based on your proposal. I agree, what I had proposed was kind of vague (and not correct the way I thought could be implemented). I have created a parent macro (for BanTSComment only) and created tests using that parent macro. Let me know your thoughts on this. We can extend it further or change it. This is just a POC.

Note: I have push the commit in a remote branch in my forked branch.

**Comment by magurotuna** [comment]
@subhakundu Your PoC looks a good first start! The next step would be:
- can it be extended to support other rules than `BanTsComment`?
- the column number is fixed per one call to `assert_lint_err_ban_ts_comment!`, and the row number cannot be specified right now. Can it be generalized so that the macro supports various error positions?

**Comment by subhakundu** [comment]
Yes, these are the next steps. Also, we need to move it to testutils.rs file. I am working on that. Thank you for your guidance.

**Comment by subhakundu** [comment]
Hi @magurotuna , I have generalized the macro to include all the parameters in this [commit](https://github.com/subhakundu/deno_lint/commit/e67f2a01d75b36788f61947bdc307d2e8a3bacea). I had to create another macro parse_err_test_2 which is almost a duplication of parse_err_test. I am working on reusing parse_err_test. Please let me know your thoughts on that. Also, I wanted to ask you a couple of questions 

1. Should we refactor all the tests in the same PR? Or we can start with ban_ts_comment.rs and no_await_in_loop.rs?
2. Is there any way, we can check the coverage? I just want to make sure, I do not decrease the coverage by chance. I found out that the number of tests passed can be a metric. Is that good?

Please let me know your thoughts.

**Comment by magurotuna** [comment]
@subhakundu 
Awesome progress.
As you said, it would be fine to reuse `parse_err_test` rather than create `parse_err_test2`.
Also, it'd be much better if `assert_lint_err_tests_generator` took another pattern, where we could specify a filename as an argument, just like `assert_lint_err` does. This is useful when adding test cases whose results depend on the target filename (e.g. `no-unused-vars` rule changes its behavior depending on the file extension)

> 1. Should we refactor all the tests in the same PR? Or we can start with ban_ts_comment.rs and no_await_in_loop.rs?

Either is fine to me, but if I'm forced to say, separate PRs are good for reviewers. 

> 2. Is there any way, we can check the coverage? I just want to make sure, I do not decrease the coverage by chance. I found out that the number of tests passed can be a metric. Is that good?

I don't think the number of tests provided by `cargo test` can be used as the coverage in this case, because the number is measured as how many test functions (i.e. functions annotated with `#[test]`) exist.  We should instead check for the number of actual test cases that are passed to the `assert_lint_xxx` macro, but I don't come up with a good way to count that...

**Comment by subhakundu** [comment]
Thank you @magurotuna, for your valuable feedback. It makes sense to create separate PRs. For now, I am working on the following things
1. Incorporating filename
2. Reusing parse_err_test (it is my highest priority as it will reuse the existing code)
3. Making changes for ban_ts_comment.rs and no_await_in_loop.rs
4. We can reuse this macro in other tests in separate PRs.

Please let me know your thoughts on this.

**Comment by subhakundu** [comment]
Hi @magurotuna, I have incorporated all the changes in the following [commit](https://github.com/subhakundu/deno_lint/commit/f989147cb72297431a14a308f65f87e24b2c5128). Please have a look at it and let me know if you see any concerns. As per point 3 in my previous comment, I have covered only two files.

Note: I have missed one unused code under a comment. I will remove it before raising the final PR.

**Comment by magurotuna** [comment]
@subhakundu Looks nice! Your latest code is almost close to adopting. I have just one comment though, which is trivial; there could be a little bit better macro name than  `assert_lint_err_tests_generator`, in my opinion. Could you consider it? I think using the same name (namely `assert_lint_err`) would also be fine, if a suitable name doesn't come to mind.

**Comment by subhakundu** [comment]
Sure, I can make that change. Thank you for your suggestions. Let me think what can be a good name, if not, I will use the same name. I will raise another PR from my forked repo. I hope that is fine. If not, please let me know the preferred way.

**Comment by subhakundu** [comment]
Hi @magurotuna, I have raised a PR for the issue. Let me know if I need to do anything else.

**Comment by subhakundu** [comment]
As the previous PR has been merged, I will cover other tests in separate PRs in the same issue. Also, if someone wants to use the macro to cover other tests, I will be more than happy to assist.
